stage3.0_first_views/traited_face_detect.py

```diff
@@ -0,0 +1,116 @@
+# General imports
+from os.path import join
+
+import PIL.Image
+import numpy as np
+from matplotlib import patches
+from matplotlib.figure import Figure
+from PIL.ExifTags import TAGS
+from skimage import data
+from skimage.feature import Cascade
+
+# ETS imports
+from traits.api import (
+    Dict,
+    File,
+    HasStrictTraits,
+    Instance,
+    List,
+    observe,
+)
+from traitsui.api import Item, ModelView, View
+
+# Local imports
+from ets_tutorial.util.mpl_figure_editor import MplFigureEditor
+
+
+class ImageFile(HasStrictTraits):
+    """ Model to hold an image file.
+    """
+    filepath = File
+
+    faces = List
+
+    metadata = Dict
+
+    def to_array(self):
+        if not self.filepath:
+            return np.array([])
+
+        with PIL.Image.open(self.filepath) as img:
+            return np.asarray(img)
+
+    @observe("filepath")
+    def _update_faces_and_metadata(self, event):
+        self.metadata = {}
+        self._update_metadata_with_exif()
+        self._detect_faces()
+
+    def _update_metadata_with_exif(self):
+        if not self.filepath:
+            return
+        with PIL.Image.open(self.filepath) as img:
+            exif = img._getexif()
+        if not exif:
+            return
+        self.metadata.update(
+            {TAGS[k]: v for k, v in exif.items() if k in TAGS}
+        )
+
+    def _detect_faces(self):
+        self.faces = []
+        if not self.filepath:
+            return
+        # Load the trained file from the module root.
+        trained_file = data.lbp_frontal_face_cascade_filename()
+        # Initialize the detector cascade.
+        detector = Cascade(trained_file)
+        self.faces = detector.detect_multi_scale(
+            img=self.to_array(),
+            scale_factor=1.2,
+            step_ratio=1,
+            min_size=(60, 60),
+            max_size=(600, 600)
+        )
+        self.metadata['Number of faces'] = len(self.faces)
+
+
+class ImageFileView(ModelView):
+    """ ModelView for an image file object.
+    """
+    model = Instance(ImageFile)
+
+    figure = Instance(Figure)
+
+    view = View(
+        Item("model.filepath", style="readonly", show_label=False),
+        Item("figure", editor=MplFigureEditor(), show_label=False),
+        resizable=True
```

Comment:
Since this code introduces quite a few concepts, already, I would propbably call this version 3.1. We could also set the title to be something like `My App` or something to show customizing this further? And maybe introduce a version 3.2 which introduces `VGroup` around the 2 items?

Reply:
Another suggestion: 3.0 could just be the model class alone, and just calling `configure_traits` on it (and only allowing the user to set the File attribute, click ok and let the rest of the script run with that choice).

Reply:
> Another suggestion: 3.0 could just be the model class alone, and just calling configure_traits on it (and only allowing the user to set the File attribute, click ok and let the rest of the script run with that choice).

Implemented in [ed112ee](https://github.com/jonathanrocher/ets_tutorial/pull/44/commits/ed112eec43eb7077fe12cd1f3877d3d3182325e9)

> Since this code introduces quite a few concepts, already, I would propbably call this version 3.1.

Renamed this version to 3.1. Using a `VGroup` made virtually no difference. How were you thinking of using it?

Reply:
It was just to be explicit about what it does and teach people about V and HGroup. But I am using them later so don't worry here.